src/scale_impls/decode.rs

```diff
@@ -42,6 +42,27 @@ where
     )
 }
 
+/// Decode data into a [`Composite`] according to a series of fields.
+/// Each field contains a type id and optionally a field name.
+pub fn decode_composite_as_fields<'resolver, R>(
+    input: &mut &[u8],
+    fields: &mut dyn FieldIter<'resolver, R::TypeId>,
+    types: &'resolver R,
+) -> Result<Composite<R::TypeId>, DecodeError>
+where
+    R: TypeResolver,
+    R::TypeId: Clone,
+{
+    // Build a Composite type to pass to a one-off visitor:
+    let mut composite = scale_decode::visitor::types::Composite::new(input, fields, types, false);
+    // Decode into a Composite value from this:
+    let val = visit_composite::<R, R::TypeId, FromMapper>(&mut composite)?;
+    // Consume remaining bytes and update input cursor:
+    composite.skip_decoding()?;
+    *input = composite.bytes_from_undecoded();
+    Ok(val)
```

Comment:
Can you use this in the `DecodeAsFields impl on line 88-96 too? (and then have the `al.map_err(From::from).map(|v| v.remove_context())` line after this call to remove the context)

Reply:
That would be nice, but it does not work yet because we are missing the necessary trait bounds in `scale-type-resolver`: The freestanding `decode_composite_as_fields` function requires `TypeId: Clone` but we cannot afford this bound in the `DecodeAsFields` impl :(

Reply:
Aah I see! Maybe this `Clone` bound on `TypeId` is another thing that should exist higher up by default (and maybe `'static` too. But for now this will do!